File: JS9Ch/src/S_6_Dynamic_Problem_II/Required_10/python_cpp/_476_Stone_Game_Medium.cpp

class Solution {
public:
    /**
     * @param A an integer array
     * @return an integer
     */
    int stoneGame(vector<int>& A) {
        // Write your code here
        int n = A.size();
        if (n <= 1)
            return 0;
        vector<int> sum, num;
        sum.resize(n+1);
        num.resize(n+1);

        vector<vector<int> > f, s;
        for (int i = 0; i <= n; ++i) {
            f.push_back(num);
            s.push_back(num);
        }
        for (int i = 1; i <= n; ++i)
            num[i] = A[i-1];

        for (int i=1; i<=n; i++) {
            sum[i] = sum[i-1] + num[i];
            f[i][i] = 0;
            s[i][i] = i;

        }
        for (int len=2; len<=n; len++)
            for (int i=1; i<=n-len+1; i++) {
                int j=i + len - 1;
                f[i][j]= 0x7fffffff / 3;
                for ( int k=s[i][j-1]; k<=s[i+1][j]; k++)
                    if (f[i][k-1]+f[k][j]+sum[j]-sum[i-1] < f[i][j]) {
                        f[i][j] = f[i][k-1]+f[k][j]+sum[j]-sum[i-1];
                        s[i][j] = k;
                }

        }
        return f[1][n];
    }
};

File: JS9Ch/src/S_6_Dynamic_Problem_II/Required_10/python_cpp/_476_Stone_Game_Medium.py

class Solution:
    # @param {int[]} A an integer array
    # @return {int} an integer
    def stoneGame(self, A):
        n = len(A)
        if n < 2:
            return 0
        dp = [[0 for _ in xrange(n)] for _ in xrange(n)]
        sum = [[0 for _ in xrange(n)] for _ in xrange(n)]

        # 0, 1, 2, ... n-2, n-1
        for i in xrange(n):
            dp[i][i] = 0
        for i in xrange(0, n-1):
            dp[i][i+1] = A[i]+A[i+1]

        for i in xrange(n):
            sum[i][i] = A[i]
            for j in xrange(i+1, n):
                sum[i][j] = sum[i][j-1] + A[j]

        for x in xrange(3, n+1):
            i = 0
            while i+x-1 < n:
                j = i+x-1
                k = i
                dp[i][j] = sys.maxint
                while k+1 <= j:
                    t = dp[i][k] + dp[k+1][j] + sum[i][k] + sum[k+1][j]
                    dp[i][j] = min(dp[i][j], t)
                    k += 1
                i += 1
        return dp[0][n-1]